src/tfr_sensor/src/light_detector.cpp

```diff
@@ -0,0 +1,85 @@
+#include <light_detector.h>
+
+/*
+ *  add an image to the moving average, drops old images if no longer needed.
+ * */
+void LightDetector::add_image(const sensor_msgs::ImageConstPtr& msg)
+{
+    //manage the moving average
+    if (brightness.size() >= width)
+        brightness.pop_back();
+    
+    //convert out of std ros image
+    cv::Mat image;
+    try
+    {
+        image = cv_bridge::toCvCopy(msg)->image;
+    }
+    catch (cv_bridge::Exception& e)
+    {
+        ROS_ERROR("cv_bridge exception: %s",
+                e.what());
+        return;
+    }
+    cv::Scalar intensities = cv::sum(image);
+
+    ColorStats stats{};
+    auto scale = [&image] (const double &intensity) 
+    {
+        return intensity/(image.rows*image.cols);
+    };
+
+    //note we have to reverse out of native cv bgr ordering
+    stats.r_ave=scale(intensities[2]);
+    stats.g_ave=scale(intensities[1]);
+    stats.b_ave=scale(intensities[0]);
+
+    brightness.push_front(stats);
+}
+
+/*
+ *  Has the light been turned on?  
+ * */
+bool LightDetector::is_on()
+{
+    //we want a full window this will take 1/20th of a second to fill
+    if (brightness.size() < width)
+        return false;
+
+    ColorStats total{};
+    int ave_size = brightness.size() -  1;
+    /*
+     *Get an average of the color channels for every element exept for the most
+     recent addition to the average.
+     * */
+    std::for_each(++brightness.begin(), brightness.end(), 
+            [&total, &ave_size] (const ColorStats &stat)
+            {
+                total.r_ave += stat.r_ave/ave_size;
+                total.g_ave += stat.g_ave/ave_size;
+                total.b_ave += stat.b_ave/ave_size;
+            });
```

Comment:
See above comment.  Normal `for` loop or even a `range for` loop would be much better for readability here.

Reply:
This is just basic functional programming, here, and uses the standard libraries effectively. This is a perfectly valid use case for a foreach.

Reply:
This is a terrible use of `for_each` because it can be replaced with a ranged-for loop without losing anything.

Reply:
well you couldn't use a ranged for, you would need to use an old fashioned for loop (with a little auto to cut down on iterator verbosity), and increment the start iterator like I do in the foreach. The issue is that we need to sum all of the elements except for the 1st element. The ranged for wouldn't expose the starting and ending iterators for me to offset.

so the alternative would be:
```
ColorStats total{};
int ave_size = brightness.size() -  1;

for(auto elem = ++brightness.begin(), auto end brightness.end(); elem != end; ++elem)
{
                total.r_ave +=  elem->r_ave/ave_size;
                total.g_ave += elem->g_ave/ave_size;
                total.b_ave += elem->b_ave/ave_size;
}
```

Reply:
You're right, you can't use ranged-for if you want to skip the first element.  I hadn't considered that.  But you don't have to use iterators here; you can just index into the data structure to get each element.

```
for (int i = 1; i < brightness.size(); i++)
    total.r_ave += brightness[i].r_ave / ave_size;
    total.g_ave += brightness[i].g_ave / ave_size;
    total.b_ave += brightness[i].b_ave / ave_size;
```

Reply:
I think I agree with @RyanBerge for this case, a for loop like he wrote is a lot more readable to me than the for_each call.

Reply:
@AdamuKaapan I think iterators are more maintainable, than indexing for areas that you don't care about performance.

Reply:
Which we don't in this method, we do in the previous one that processes the video stream